### PR TITLE
[graph_trainer] in-graph inplace gradient accumulation mode, to make it CudaGraph friendly

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Option to include grad accumulation inside the graph and do inplace accumulation, which is CudaGraphable.

For this GradState is extracted to a dataclass,
Grad accumulation is included into a graph as add_ node.

The list of functionality is not supported yet and needs additional testing - there is extensive validation:
 - no parameters aliasing (tight params)
 - no PP
 - no Precompile

Authored with assistance from an AI coding assistant.

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/gradient_accumulation.py [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+        cls._validate_unique_parameters(named_parameters)
+        parameters = tuple(parameter for _, parameter in named_parameters)
+        cls._validate_optimizer_params_membership(parameters, optimizers)
+
+        buffers = tuple(torch.zeros_like(parameter) for parameter in parameters)
+        state = cls(
+            parameter_fqns=tuple(fqn for fqn, _ in named_parameters),
+            parameters=parameters,
+            parameter_storage_keys=tuple(
+                _storage_keys(parameter) for parameter in parameters
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/gradient_accumulation.py:96]: nit: This iteration can be avoided by doing it once before _validate_unique_parameters() and passing as argument to _validate_unique_parameters

File: torchtitan/experiments/graph_trainer/trainer.py [hunk 2]
```diff
@@ -336,6 +424,17 @@ def train_step(
         PRE_TRAIN_STEP_HOOKS.get(self.config.compile.pass_pipeline, lambda _: None)(
             self
         )
+        if self.config.compile.enable_inplace_graph_gradient_accumulation:
+            if len(self.model_parts) != 1:
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/trainer.py:428]: This check can be done earlier and not every train_step

File: torchtitan/experiments/graph_trainer/trainer.py [hunk 3]
```diff
@@ -336,6 +424,17 @@ def train_step(
...
+                raise RuntimeError("in-graph gradient accumulation requires one model")
+            model = self.model_parts[0]
+            gradient_state = self._ensure_graph_gradient_state(model)
+            parameters = tuple(
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/trainer.py:432]: This feels like an expensive check for every train_step. Can we really expect parameter identities or buffer mappings to change from step to step? Can we do it once?

File: torchtitan/experiments/graph_trainer/gradient_accumulation.py [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+    for key in ("custom", "unbacked_bindings"):
+        if isinstance(value := sink.meta.get(key), dict):
+            sink.meta[key] = copy.copy(value)
+    if "val" in buffer.meta:
+        sink.meta["val"] = buffer.meta["val"]
+    sink.meta["graph_gradient_fqn"] = fqn
+    return sink
+
+
+def finalize_graph_gradient_accumulation(
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/gradient_accumulation.py:407]: Hmm, why is this needed? Can't we directly have the minimal_fx_tracer to trace through the gradient accumulation operations as well?

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 5]
```diff
@@ -285,6 +285,82 @@ def _reparametrize_train_state(
         yield
 
 
+@dataclass(frozen=True, slots=True)
+class GraphStateMapping:
+    """Trace-time locations associated with one graph-owned tensor.
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py:290]: nit: "graph-owned" is not an accurate name for this. More like "trainer-owned, but graph-mutated" ?

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py [hunk 6]
```diff
@@ -18,18 +18,29 @@
 )
 from torchtitan.experiments.graph_trainer.common_utils import (
     _maybe_materialize_grad_for_param_layout,
+    accumulate_param_grads_,
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/tests/test_trace_module.py:21]: According to an agent review of testing completeness, there are two gaps worth fixing:

## Main gaps

1. The CUDA test does not explicitly prove capture occurred.

[test_cuda_graph_numerics_match_external_accumulation()](command:code-compose.open?%5B%22%2Fhome%2Favenkataraman%2Fgithub%2Ftorchtitan%2Ftorchtitan%2Fexperiments%2Fgraph_trainer%2Ftests%2Ftest_trace_module.py%22%2C165%5D) relies on the current one-iteration warmup. It checks that the wrapper exists, but not:

self.assertIsNotNone(wrapper._graph)
Without this, increasing the warmup count could cause the test to pass without testing capture/replay.

2. CUDA Graph + FSDP/DTensor is not tested together.

The CUDA test uses an ordinary nn.Linear. The FSDP test validates DTensor accumulation but executes the FX graph directly. Since most of the complexity in finalize_graph_gradient_accumulation() concerns flattened DTensor buffers, this combination deserves one test.

A useful strengthening of the CUDA test would also assert that gradient-buffer inputs are included in _static_input_indices and record their pointers before warmup, then check them after capture, replay, zeroing, and the next step.

A full public train_step() test and parameterized unsupported-config tests would be beneficial, but I would consider them secondary. I would prioritize the explicit-capture assertion and one FSDP CUDA Graph test.

**Review by aditvenk — COMMENTED** [review]
LGTM, big question is around the need for the finalize_graph_gradient_accumulation pass (which is the most complex change in this PR) -- and whether we can avoid it by directly having the tracer to trace the gradient accumulation operations into the graph.

File: torchtitan/experiments/graph_trainer/trainer.py [hunk 7]
```diff
@@ -280,6 +299,11 @@ def _make_fx_forward_backward_step(
                 self._traced_step.gm = remove_parameter_gradient_markers_pass(
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/trainer.py:299]: What does it mean for `self.config.compile.enable_passes` to be `False` , but we are still running these two passes? 

Is there some "ABI"-type reasons for running these passes unconditionaly? If yes, should there be an "construct_mandatory_graph_passes" pipeline that we can construct with the necessary passes? We can then simplify the code in the trainer to construct either the default pipeline or the mandatory pipeline depending on value of enable_passes
